mlflow/transformers/__init__.py

```diff
@@ -2761,6 +2802,32 @@ def is_uri(s):
                 error_code=BAD_REQUEST,
             )
 
+    def _wrap_input_in_prompt_template(self, input_data):
+        """
+        Wraps the input data in the specified prompt template. If no template is
+        specified, or if the pipeline is an unsupported type, or if the input type
+        is not a string or list of strings, then the input data is returned unchanged.
+        """
+
+        # if no template is set, do nothing
+        if not self.prompt_template:
+            return input_data
+
+        if self.pipeline.task not in _SUPPORTED_PROMPT_TEMPLATING_TASK_TYPES:
+            return input_data
+
+        if isinstance(input_data, str):
+            return self.prompt_template.format(prompt=input_data)
+        elif isinstance(input_data, list):
+            # if every item is a string, then apply formatting to every item
+            if all(isinstance(data, str) for data in input_data):
+                return [self.prompt_template.format(prompt=data) for data in input_data]
+            # otherwise, the input might be more complex (e.g. a dict), so we leave it alone
```

Comment:
Since we use the string formatting, is it possible to support simple Dict[str, str] like this?
```
self.prompt_template.format(**input_data)
```
I think this makes the template much more expressive

Reply:
it is possible, the only thing is that the data passed into this function should not be a dict for the pipelines that we support. there are some pre-processing steps above that make the input data into `str` / `List[str]` ([example](https://github.com/mlflow/mlflow/blob/438a450714a3ca06285eeea34bdc6cf79d7f6cbc/mlflow/transformers/__init__.py#L2415))

Reply:
Oh, then why don't we do template formatting before `_parse_text2text_input` or even replacing it? The function appears to sim similar goal as prompt templating (with much more limited way).

Reply:
+1 to @B-Step62's suggestion :)

Reply:
I did some investigation on this, and while it is possible, I worry that it comes at the cost of ease-of-use for users. Here's what I think, let me know if it makes sense. For now, I'm leaning toward leaving it as it is, but I'm open to everyone's thoughts:

### Restricting the prompt template to have a single {prompt} placeholder:
**Pros:**
1. less can go wrong. if it works at validation time, it will work at runtime with no work required by users.
2. easy to educate about the feature. since there is only one format allowed, error messages and tutorials can clearly specify what is required.
3. the door is still open to supporting more keys (or arbitrary keys) in the future, by simply relaxing our validation function

**Cons:**
1. Less expressive
  a. I'm not sure how much this actually comes into play though. The text2text-generation expects a string input, and though we support passing in arbitrary dicts here, the default model signature for this pipeline is string, and it's also what we show in our official examples. Users can of course specify their own signature to get around this, but I don't know how common this is in practice.

### Allowing user to specify their own variables in the template:
**Pros:**
1. More expressive

**Cons:**
1. Prompt template can fail at runtime if user doesn't ensure their their data matches up with their input
  a. Since validation will be looser to support this feature, it's possible that regular users can save prompt templates that won't work by accident (e.g. having an extra placeholder)
  b. Error messages for this will be slightly harder to craft, because we won't know what the user intended—is their prompt template correct but data wrong? or is their prompt template wrong but data correct? It's not an impossible problem to solve (we can use some heuristics like if the prompt template has keys, then probably the the data is wrong), but I don't think it's trivial to create a great user experience for this.
2. More things can go wrong, meaning more maintenance load for us.
  a. Despite the fact that most users probably won't run into this problem, we still have to handle these situations gracefully. More helpful error messages to craft, more documentation to write, etc.
  b. This is more work upfront (not a lot but not insignificant), and also means we'll be changing how text2text input parsing works. Given that we're already implementing this new feature, i think it's less risky to minimize changes to existing behavior at least for the moment.

Reply:
Will all of the intended-to-support pipeline types work with a single input prompt within a template? Can you do a quick validation to ensure that compound prompt elements are not needed? If that's the case, restricting to a single named key makes sense to me and I am in favor of reducing the complexity for this as much as possible. 
If there are targeted pipelines where multiple interpolation keys are needed or are directly supported as a key feature of the pipeline type, we should just replace the logic within the existing parser to incorporate the template there, then provide guidance in the docs that explains how to validate which elements from a prompt template are required to be supplied.

Reply:
Thanks for the investigation, Daniel!
I'm fine with keeping it as it is. I re-read [the original request](https://docs.google.com/document/d/1e8PTbnqWqS7gU0FW1nBpxzWeguiXfqzhXLms7kZWlRU/edit) more carefully, and only supporting `prompt` is enough. I imagined more general use cases like [Langchain](https://python.langchain.com/docs/modules/model_io/prompts/few_shot_examples) and [Microsoft](https://learn.microsoft.com/en-us/semantic-kernel/prompts/prompt-template-syntax), but it seems the target scenario is limited now.

Regardless of which, let me just clarify a few points about complexity, in case we look back on this thread for new requests. No change needed now.
1. For validation and error message, I think Daniel's implementation of `_validate_prompt_template` already has all what we need. With that we can get list of placeholders in the saved template, compare with the input keys, and say "You have different key sets for prompt templating blah blah". 
 ```
    format_args = [
        tup[1] for tup in string.Formatter().parse(prompt_template) if tup[1] is not None
    ]
 ```
2.  In terms of what users can input at inference time, **`_parse_text2text_input` is more permissive** than kv-based string fomatting. Regardless of prompt template, user can input any keys as data - like both `{"foo": "foo"} and {"bar": "bar"}`are valid input - and tbh I'm not sure we want to keep this type of use case. Imo if we want to support Dict input, we should template it, otherwise limiting to `str` or `List[str]` makes more sense to me.

Reply:
@BenWilson2 yup, the 5 pipelines supported here are all `str` input pipelines. when we extend this in the future for `question-answering` pipelines, then compound prompt elements will be necessary, but for now it isn't.

@B-Step62
1. Yup! I think the current implementation is flexible enough that we can extend in the future for more complex use-cases if we need to.
2. Agree on this—I think that if the user creates a text2text pipeline, they should pass the input as string. It's a bit confusing that we allow an arbitrary dict and just concat the inputs, though maybe there's some historical reason that we're missing context on.